File: LeetCode/Hard/Count Subarrays With Fixed Bounds/Go/countSubarraysWithFixedBounds.go
package main

import (
	"fmt"
	"math"
)

//using two pointers approach
func countSubarrays(nums []int, minK int, maxK int) int64 {
	size := len(nums)
	subarray := int64(0)
	minPos := -1
	maxPos := -1
	left := -1

	for i := 0; i < size; i++ {
		if nums[i] < minK || nums[i] > maxK {
			left = i
		}
		if nums[i] == minK {
			minPos = i
		}
		if nums[i] == maxK {
			maxPos = i
		}
		subarray += int64(max(0, min(maxPos, minPos)-left))
	}
	return subarray
}

func countSubarraysBruteForce(nums []int, minK int, maxK int) int64 {
	size := len(nums)

	subarray := int64(0)
	for i := 0; i < size; i++ {
		minVal := math.MaxInt64
		maxVal := math.MinInt64
		for j := i; j < size; j++ {
			minVal = min(minVal, nums[j])
			maxVal = max(maxVal, nums[j])
			if minVal == minK && maxVal == maxK {
				subarray++
			}
		}
	}

	return subarray
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(countSubarrays([]int{1, 3, 5, 2, 7, 5}, 1, 5))
}

File: LeetCode/Hard/Count Subarrays With Fixed Bounds/Java/countSubarraysWithFixedBounds.java
class Solution {
    public long countSubarrays(int[] nums, int minK, int maxK) {
        int size = nums.length;
        int left = -1;
        int minPos = -1;
        int maxPos = -1;
        long subarrays = 0;

        for (int i=0; i<size; i++) {
            if (nums[i]<minK || nums[i]>maxK) left = i;
            if (nums[i] == minK) minPos = i;
            if (nums[i] == maxK) maxPos = i;
            subarrays += Math.max(0, Math.min(minPos, maxPos)-left);
        }

        return subarrays;
    }
}
